### ISSUE TITLE
tasks: shrink size or/and number of task_manager tasks

### ISSUE THREAD

**Issue by Deexie** [scylladb/scylladb#16694]
Task manager tasks are kept in task manager for task_ttl (10s by default) 
after they are finished. When operations tracked by task manager are 
invoked extensively (e.g. when we add and delete a large number of tables), 
many tasks (which often represent empty operations) are kept in memory, 
leading to OOM.

Size or/and number of tasks kept in memory needs to be reduced.

See: #16493.

**Comment by mykaul** [comment]
6.0? 6.1?

**Comment by bhalevy** [comment]
I think we can target 6.0
and plan on backport to 5.4

so avoid intrusive changes if possible and defer
changes in design (like folding suboridinate tasks status to parent upon completion, so they can be destroyed on done()) to be done as followup.

**Comment by bhalevy** [comment]
Related issues:
https://github.com/scylladb/scylladb/issues/16803
https://github.com/scylladb/scylladb/issues/16804

**Comment by mykaul** [comment]
Is this fixed? I remember there were multiple fixes (one in Seastar, the other in Scylla, etc.) ?

**Comment by Deexie** [comment]
> Is this fixed? I remember there were multiple fixes (one in Seastar, the other in Scylla, etc.) ?

Related issues were fixed, I think so is this.

There were also other ideas to modify tasks, but they were rather minor and not urgent, so I think they can be standalone.

**Comment by denesb** [comment]
Closing this issue per the above.

**Comment by denesb** [comment]
We have backported https://github.com/scylladb/scylladb/commit/697ebef1490ecc1e3351af825ab56d5c973fa366 (the main fix) to 5.4, dropping labels.

**Comment by tgrabiec** [comment]
Apparently it is not fixed, it's causing problems with compaction and tablets: https://github.com/scylladb/scylladb/issues/18135#issuecomment-2077284020

**Comment by tgrabiec** [comment]
Why do we keep tasks around for 10s? Seems arbitrary.

**Comment by denesb** [comment]
> Why do we keep tasks around for 10s? Seems arbitrary.

So user has a window to query the state of finished tasks. Otherwise the moment a task is finished, it is erased from the records.

**Comment by raphaelsc** [comment]
> > Why do we keep tasks around for 10s? Seems arbitrary.
> 
> So user has a window to query the state of finished tasks. Otherwise the moment a task is finished, it is erased from the records.

shouldn't we have a limit on number of finished tasks? that is to avoid surprises in production. we thought we fixed it, but then there were another place that triggered it. what if we miss another one? I suggest we enforce such a limit.

**Comment by bhalevy** [comment]
> > > Why do we keep tasks around for 10s? Seems arbitrary.
> > 
> > 
> > So user has a window to query the state of finished tasks. Otherwise the moment a task is finished, it is erased from the records.
> 
> shouldn't we have a limit on number of finished tasks? that is to avoid surprises in production. we thought we fixed it, but then there were another condition that triggered it. what if we miss another one? I suggest we enforce such a limit.

I don't really like that.
Maybe if we distinguished between user-initiated task vs. internal tasks and keep only the former around for interactive usage or automation scripts we would prevent accumulation of large volumes of internal tasks, assuming that a user or a script won't overflow the system and would wait on async tasks it created, and then they would be destroyed as soon as their done status is retrieved over the api.

**Comment by raphaelsc** [comment]
"assuming that a user or a script won't overflow the system and would wait on async tasks it created". I think this is not a good assumption to have. A bug in a script or whatever could cause node to go unresponsive. If the task had something to do and we have serialization of maintenance tasks in some layer, then we have protection. But we'd still be susceptible to the task accumulation problem if user-initiated tasks had nothing to do, as they can finish right away and will still remain alive for 10s.

**Comment by bhalevy** [comment]
> "assuming that a user or a script won't overflow the system and would wait on async tasks it created". I think this is not a good assumption to have. A bug in a script or whatever could cause node to go unresponsive. If the task had something to do and we have serialization of maintenance tasks in some layer, then we have protection. But we'd still be susceptible to the task accumulation problem if user-initiated tasks had nothing to do, as they can finish right away and will still remain alive for 10s.

We can refuse or block user-initiated tasks if there are too many such tasks queued up as a form of load shedding / back-pressure.

**Comment by raphaelsc** [comment]
> > "assuming that a user or a script won't overflow the system and would wait on async tasks it created". I think this is not a good assumption to have. A bug in a script or whatever could cause node to go unresponsive. If the task had something to do and we have serialization of maintenance tasks in some layer, then we have protection. But we'd still be susceptible to the task accumulation problem if user-initiated tasks had nothing to do, as they can finish right away and will still remain alive for 10s.
> 
> We can refuse or block user-initiated tasks if there are too many such tasks queued up as a form of load shedding / back-pressure.

But should we prevent new tasks while taking finished tasks into account? refuse new tasks happen on ongoing tasks, not finished ones.

**Comment by raphaelsc** [comment]
think for example, user is constantly submitting new tasks that had nothing to do, one after the other. how many can accumulate in 10s period?

**Comment by bhalevy** [comment]
> > > "assuming that a user or a script won't overflow the system and would wait on async tasks it created". I think this is not a good assumption to have. A bug in a script or whatever could cause node to go unresponsive. If the task had something to do and we have serialization of maintenance tasks in some layer, then we have protection. But we'd still be susceptible to the task accumulation problem if user-initiated tasks had nothing to do, as they can finish right away and will still remain alive for 10s.
> > 
> > 
> > We can refuse or block user-initiated tasks if there are too many such tasks queued up as a form of load shedding / back-pressure.
> 
> But should we prevent new tasks while taking finished tasks into account? refuse new tasks happen on ongoing tasks, not finished ones.

My comment assumes a model where only user-initiated tasks wait for their initiator to grab their status when done, therefore it makes sense to refuse new tasks if the user creates too many tasks without waiting on the previous ones to complete.

**Comment by raphaelsc** [comment]
> > > > "assuming that a user or a script won't overflow the system and would wait on async tasks it created". I think this is not a good assumption to have. A bug in a script or whatever could cause node to go unresponsive. If the task had something to do and we have serialization of maintenance tasks in some layer, then we have protection. But we'd still be susceptible to the task accumulation problem if user-initiated tasks had nothing to do, as they can finish right away and will still remain alive for 10s.
> > > 
> > > 
> > > We can refuse or block user-initiated tasks if there are too many such tasks queued up as a form of load shedding / back-pressure.
> > 
> > 
> > But should we prevent new tasks while taking finished tasks into account? refuse new tasks happen on ongoing tasks, not finished ones.
> 
> My comment assumes a model where only user-initiated tasks wait for their initiator to grab their status when done, therefore it makes sense to refuse new tasks if the user creates too many tasks without waiting on the previous ones to complete.

but why do we add 10s delay to the user?

**Comment by raphaelsc** [comment]
if we don't wait for task + 10s delay, then problem is still there.

**Comment by bhalevy** [comment]
> if we don't wait for task + 10s delay, then problem is still there.

Right.  The user is expected to wait on the task in order to get its completion status.
The reason we keep it around is to be able to distinguish between task IDs that never existed from those who finished and are not tracked anymore.

For the legacy repair async api, tasks were identified by a sequence number so we could tell if a task id is already done based on the last id we dispatched.  We kept around only error statuses, forever, until the user picked them up.
But with UUID based task-id:s (that are such on purpose to facilitate distributed creation of them), we can't do much with an unknown task-id.
We can decide to keep only failed tasks around for the user to fetch their status, but we may end up with the same problem if we have a long batch of failed tasks.

**Comment by Deexie** [comment]
> We can refuse or block user-initiated tasks if there are too many such tasks queued up as a form of load shedding / back-pressure.

I'm not sure if it's okey to allow only one user requested operation of the same type at the time. 

If it were we could:
- add similar mechanism to `retrier` from `storage_service::raft_topology_cmd_handler` (https://github.com/scylladb/scylladb/blob/5ca9a46a917a47b5c2fad8b35e8e6f64b1cd8b2c/service/storage_service.cc#L5210-L5228) i.e. if user requests e.g. repair while other user requested repair is running, it gets the task_id of old (currently running) repair;
- if one operation finished and user starts the other one then the task (tasks tree) for the old operation is removed from task manager.

That would require API calls which start the operation to be asynchronous. Synchronous one would need to throw if another operation is running to prevent accumulation of waiters. 


If we rather need to allow reasonable number (say `n`) of the same user requested operation, then we could count existing tasks and if there is `n` tasks:
- check if some of them are done	
	- if so - remove them and run new operation;
	- otherwise - throw.

That would require a lot of blocking though. I don't really like this idea, but I see no other solution.

**Comment by Deexie** [comment]
> I suggested earlier we introduce a notion of did_work and only keep alive tasks that had something to do

At first I really liked `did_work` idea but on the second thought:
let's say we have an operation that has node, keyspace, and table level tasks. Assume that half of table-level tasks has nothing to do. We mark them as "did no work" and remove them from task manager. But parents task (keyspace-level) of these tasks still keep foreign pointers to them. If we want to delete them then each task needs:
- a lock for children vector as we no longer have the assumption that its size is growing monotonously;
- some (?) variables for including the immediately removed children in progress of their parents.

We would also need to add some additional checks like "if keyspace-level task finished and has no children then remove it immediately". It really complicates the code.

I think that we should check if a task has anything to do before it is created and if check is too long, too complicated or not possible beforehand, then maybe it do is worth the task. Maybe it's worth revising if currently implemented tasks do so.

**Comment by raphaelsc** [comment]
> > I suggested earlier we introduce a notion of did_work and only keep alive tasks that had something to do
> 
> At first I really liked `did_work` idea but on the second thought: let's say we have an operation that has node, keyspace, and table level tasks. Assume that half of table-level tasks has nothing to do. We mark them as "did no work" and remove them from task manager. But parents task (keyspace-level) of these tasks still keep foreign pointers to them. If we want to delete them then each task needs:
> 
> * a lock for children vector as we no longer have the assumption that its size is growing monotonously;
> * some (?) variables for including the immediately removed children in progress of their parents.
> 
> We would also need to add some additional checks like "if keyspace-level task finished and has no children then remove it immediately". It really complicates the code.
> 
> I think that we should check if a task has anything to do before it is created and if check is too long, too complicated or not possible beforehand, then maybe it do is worth the task. Maybe it's worth revising if currently implemented tasks do so.

I think we can do it but we still need some protection elsewhere (maybe something like @bhalevy suggested above). Someone can add a new task type tomorrow, forgets to do it because it's not enforced in the interface, as it would with did_work, then we fall into the same problem. We're targeting 100 tablets per shard, maybe after your series to allocate groups dynamically, we will not have this problem anymore caught by longevity. The priority decreases significantly after your series is merged,  I think.

**Comment by bhalevy** [comment]
> Assume that half of table-level tasks has nothing to do. We mark them as "did no work" and remove them from task manager. But parents task (keyspace-level) of these tasks still keep foreign pointers to them. If we want to delete them then each task needs:
> 
> * a lock for children vector as we no longer have the assumption that its size is growing monotonously;
> * some (?) variables for including the immediately removed children in progress of their parents.

It sounds complicated but I think it's worth considering folding the completion status of children tasks into their parent when they are done and then deallocating the task objects so we can collapse the task tree as it accomplishes its goals, until we're left only with the top-level task, when it's not internal (i.e. it's trackable over the api).  Otherwise, if it's internal, it can also evaporate immediately when it's done (recursively, after all its children tasks are done).
